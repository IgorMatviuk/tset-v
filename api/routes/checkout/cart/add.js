const _cloneDeep = require('lodash.clonedeep')
const shortid = require('shortid')

const { products } = require('../../../factories')
const cart = require('../../../fixtures/cart')

const byKey = (key, value) => (item) => item[key] === value

function add(req, res) {
  const id = req.params.id

  const productById = byKey('variant_id', id)
  const existedProduct = cart.items.find(productById)

  if (existedProduct) {
    existedProduct.quantity += 1
  } else {
    const targetProductById = byKey('id', id)
    const targetProduct = products.find((product) =>
      product.variants.find(targetProductById)
    )
    const cloneProduct = _cloneDeep(targetProduct)
    const nextItemId = shortid.generate()

    cart.items.push({
      id: nextItemId,
      variant_id: id,
      product: { ...cloneProduct },
      quantity: 1
    })
  }

  const total = cart.items.reduce((total, item) => {
    const variant = item.product.variants.find((x) => x.id === item.variant_id)
    const price = variant.discount_price || variant.price
    total += price * item.quantity
    return total
  }, 0)

  const itemsQty = cart.items.length

  cart.grand_total = total.toFixed(4)
  cart.formatted_grand_total = `${total} грн`
  cart.items_qty = itemsQty.toFixed(4)

  res.jsonp({
    data: cart
  })
}

module.exports = add
