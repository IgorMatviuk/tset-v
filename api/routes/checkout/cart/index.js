const cart = require('../../../fixtures/cart')

function checkout(req, res) {
  res.jsonp({
    data: cart
  })
}

module.exports = checkout
