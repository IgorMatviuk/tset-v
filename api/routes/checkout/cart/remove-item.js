const cart = require('../../../fixtures/cart')

function removeItem(req, res) {
  const itemIndex = cart.items.findIndex((item) => item.id === req.params.id)

  if (itemIndex > -1) {
    cart.items.splice(itemIndex, 1)

    const total = cart.items.reduce((total, item) => {
      const variant = item.product.variants.find(
        (x) => x.id === item.variant_id
      )
      const price = variant.discount_price || variant.price
      total += price * item.quantity
      return total
    }, 0)

    const itemsQty = cart.items.length

    cart.grand_total = total.toFixed(4)
    cart.formatted_grand_total = `${total} грн`
    cart.items_qty = itemsQty.toFixed(4)
  }

  res.jsonp({
    data: cart
  })
}

module.exports = removeItem
