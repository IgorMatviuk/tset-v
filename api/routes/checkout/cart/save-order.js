function saveOrder(req, res) {
  res.jsonp({
    message: 'The order has been successfully saved'
  })
}

module.exports = saveOrder
