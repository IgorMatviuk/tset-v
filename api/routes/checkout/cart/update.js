const cart = require('../../../fixtures/cart')

const byId = (id) => (item) => item.id === id

function update(req, res) {
  // eslint-disable-next-line
  const { item_id } = req.params
  const { variant, quantity } = req.body

  const item = cart.items.find(byId(Number(item_id)))

  if (variant) {
    item.variant_id = variant
  }

  item.quantity = quantity

  const total = cart.items.reduce((total, item) => {
    const variant = item.product.variants.find((x) => x.id === item.variant_id)
    const price = variant.discount_price || variant.price
    total += price * item.quantity
    return total
  }, 0)

  const itemsQty = cart.items.length

  cart.grand_total = total.toFixed(4)
  cart.formatted_grand_total = `${total} грн`
  cart.items_qty = itemsQty.toFixed(4)

  res.jsonp({
    data: cart
  })
}

module.exports = update
