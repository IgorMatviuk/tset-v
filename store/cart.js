import { mutationTypes as localStorageMutationTypes } from '@/store/localStorage'

export const mutationTypes = {
  SET_LOADING_START: 'SET_LOADING_START',
  SET_LOADING_SUCCESS: 'SET_LOADING_SUCCESS',
  SET_CART_DATA: 'SET_CART_DATA',
  SET_COUNTRIES: 'SET_COUNTRIES',
  SET_REGIONS: 'SET_REGIONS',
  SET_STATES: 'SET_STATES',
  SET_ORDER_SAVE: 'SET_ORDER_SAVE',
  CLEAR_CART: 'CLEAR_CART'
}

export const actionTypes = {
  GET_CHECKOUT_DATA: 'GET_CHECKOUT_DATA',
  GET_CART: 'GET_CART',
  GET_COUNTRIES: 'GET_COUNTRIES',
  GET_REGIONS: 'GET_REGIONS',
  GET_STATES: 'GET_STATES',
  GET_OFFICES: 'GET_OFFICES',
  ADD_ITEM: 'ADD_ITEM',
  UPDATE_ITEM: 'UPDATE_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM'
}

export const state = () => ({
  cartData: {},
  countries: [],
  regions: [],
  states: [],
  offices: [],
  isOrderSaved: false,
  isLoading: false
})

export const mutations = {
  [mutationTypes.SET_CART_DATA](state, cartData) {
    state.cartData = cartData
  },
  [mutationTypes.SET_LOADING_START](state) {
    state.isLoading = true
  },
  [mutationTypes.SET_LOADING_SUCCESS](state, { key, value }) {
    if (state.isLoading) {
      state.isLoading = false
    }

    state[key] = value
  },
  [mutationTypes.SET_COUNTRIES](state, countries) {
    state.countries = countries
  },
  [mutationTypes.SET_REGIONS](state, regions) {
    state.regions = regions
  },
  [mutationTypes.SET_STATES](state, states) {
    state.states = states
  },
  [mutationTypes.SET_OFFICES](state, offices) {
    state.offices = offices
  },
  [mutationTypes.SET_ORDER_SAVE](state) {
    state.isOrderSaved = true
    state.cartData = {}
  },
  [mutationTypes.CLEAR_CART](state) {
    state.cartData = {}
    state.isOrderSaved = false
  }
}

export const actions = {
  async [actionTypes.GET_CHECKOUT_DATA]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: checkoutData } = await this.$axios.$get(
      `${STAGE_API_URL}/api/checkout`
    )

    commit(mutationTypes.SET_COUNTRIES, checkoutData.countries)
    commit(mutationTypes.SET_REGIONS, checkoutData.regions)
  },
  async [actionTypes.GET_CART]({ commit }, cartId) {
    // const { data: cartData } = await this.$axios.$get('/api/checkout/cart')

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    // const STAGE_API_URL = 'http://localhost:7000'
    // const { data: cartData } = await this.$axios.$get(
    //   `${STAGE_API_URL}/api/checkout/cart`
    // )

    const { data: cartData } = await this.$axios.$get(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}`
    )

    if (cartData) {
      commit(mutationTypes.SET_CART_DATA, cartData)
    }
  },
  async [actionTypes.GET_REGIONS]({ commit }, regionId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: regions } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/regions`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'regions',
      value: regions
    })
  },
  async [actionTypes.GET_STATES]({ commit }, regionId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: states } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/states?region_id=${regionId}`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'states',
      value: states
    })
  },
  async [actionTypes.GET_OFFICES]({ commit }, stateId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: offices } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/offices?state_id=${stateId}`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'offices',
      value: offices
    })
  },
  async [actionTypes.ADD_ITEM]({ commit }, { cartId, itemId }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    // const { data: cartData } = await this.$axios.$post(
    //   `/api/checkout/cart/add/${id}`
    // )

    const STAGE_API_URL = 'https://back.virna.com.ua'
    // const STAGE_API_URL = 'http://localhost:7000'
    const { data: cartData } = await this.$axios.$post(
      `${STAGE_API_URL}/api/checkout/cart/add`,
      {
        cart: cartId,
        variant: itemId
      }
    )

    if (!state.cartData) {
      commit(
        `localStorage/${localStorageMutationTypes.SET_CART_ID}`,
        cartData.id,
        { root: true }
      )
    }

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData
    })
  },
  async [actionTypes.UPDATE_ITEM]({ commit }, { cartId, itemId, body }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: cartData } = await this.$axios.$put(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}/update/${itemId}`,
      body
    )

    // const { data: cartData } = await this.$axios.$put(
    //   `/api/checkout/cart/update/${id}`,
    //   body
    // )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData
    })
  },
  async [actionTypes.REMOVE_ITEM]({ commit }, { cartId, itemId }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: cartData } = await this.$axios.$delete(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}/remove-item/${itemId}`
    )

    if (!cartData) {
      commit(`localStorage/${localStorageMutationTypes.SET_CART_ID}`, null, {
        root: true
      })
    }

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData || {}
    })
  }
}
