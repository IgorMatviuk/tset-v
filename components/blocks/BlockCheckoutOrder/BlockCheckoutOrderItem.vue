<template>
  <div class="order-item">
    <button class="order-item__remove" type="button" @click="removeItem">
      <svg-icon name="trash" class="order-item__icon" width="15" height="15" />
    </button>
    <div class="order-item__image-wrapper">
      <ImageLoader>
        <picture>
          <source
            :data-srcset="variant.base_image.thumb_xs | webp"
            srcset="/img/stub.png"
            type="image/webp"
          />
          <img
            :data-src="variant.base_image.thumb_xs"
            src="/img/stub.png"
            :alt="variant.name"
            :title="variant.name"
            class="order-item__image lazyload"
            loading="lazy"
            data-object-fit="cover"
          />
        </picture>
      </ImageLoader>
    </div>
    <div class="order-item__body">
      <NuxtLink :to="$i18n.path(variant.url_key)" class="order-item__name">
        {{ variant.name }}
      </NuxtLink>
      <div class="order-item__price-wrapper">
        <template v-if="variant.discount_price">
          <div class="order-item__price">
            {{ variant.formatted_discount_price }}
          </div>
          <div class="order-item__old-price">
            {{ variant.formatted_price }}
          </div>
        </template>
        <template v-else>
          <div class="order-item__price">
            {{ variant.formatted_price }}
          </div>
        </template>
      </div>
      <div class="order-item__code">
        {{ $t('productPage.sku') }}: {{ variant.sku }}
      </div>
      <!-- <OptionSelect
        :id="getOptionsKey('color', item.id)"
        :attribute-id="colorAttribute.id"
        :attribute-code="colorAttribute.code"
        :active-option="colorAttribute.selected_option"
        :options="mappedColorOptionItems"
        class="order-item__colors"
        @change="onOptionsChange"
      /> -->
      <OptionSelect
        :id="getOptionsKey('size', item.id)"
        :attribute-id="sizeAttribute.id"
        :attribute-code="sizeAttribute.code"
        :active-option="getActiveOption('size')"
        :options="mappedSizeOptionItems"
        class="order-item__sizes"
        @change="onOptionsChange"
      />
      <QuantityCounter
        :count="item.quantity"
        class="order-item__counter"
        @change="onQuantityChange"
      />
    </div>
  </div>
</template>

<script>
import OptionSelect from '@/components/shared/OptionSelect'
import QuantityCounter from '@/components/shared/QuantityCounter'
import ImageLoader from '@/components/shared/ImageLoader'

import { actionTypes as cartActionTypes } from '@/store/cart'

import productVariants from '@/mixins/product-variants'
import removeFromCart from '@/mixins/remove-from-cart'

import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    OptionSelect,
    QuantityCounter,
    ImageLoader
  },
  mixins: [productVariants, removeFromCart, imageNextGen],
  props: {
    index: {
      type: Number,
      required: true
    },
    item: {
      type: Object,
      required: true
    }
  },
  computed: {
    variant() {
      return this.item.product.variants.find(
        (variant) => variant.id === this.item.variant_id
      )
    },
    // colorAttribute() {
    //   return this.item.product.attributes.find((attr) => attr.code === 'color')
    // },
    sizeAttribute() {
      return this.item.product.attributes.find((attr) => attr.code === 'size')
    },
    // mappedColorOptionItems() {
    //   return this.mapOptionsWithStatus('color')
    // },
    mappedSizeOptionItems() {
      return this.mapOptionsWithStatus('size')
    }
  },
  methods: {
    mapOptionsWithStatus(attrCode) {
      let options
      let activeOppositeVariants

      /* istanbul ignore else */
      // if (attrCode === 'color') {
      //   options = this.colorOptions
      //   activeOppositeVariants = this.activeSizeVariants
      // }

      /* istanbul ignore else */
      if (attrCode === 'size') {
        options = this.sizeOptions
        activeOppositeVariants = this.activeColorVariants
      }

      return options.items.map((item) => {
        const intersection = item.variants.filter((variant) =>
          activeOppositeVariants.includes(variant)
        )

        return {
          disabled: !intersection.length,
          ...item
        }
      })
    },
    getOptionsKey(scope, id) {
      const capitalized = scope.charAt(0).toUpperCase() + scope.slice(1)
      return `order${capitalized}_${id}_${this.index}`
    },
    getActiveOption(attrCode) {
      const pAttrs = this.item.product.attributes
      const pVariants = this.item.product.variants

      const variant = pVariants.find((x) => x.id === this.item.variant_id)
      const vAttrs = variant.attributes

      const sourceAttr = pAttrs.find((x) => x.code === attrCode)
      const { value: optionId } = vAttrs.find((x) => x.code === attrCode)

      return sourceAttr.options.find((x) => x.id === optionId)
    },
    getIntersectedVariantId() {
      return this.activeColorVariants.find((cVariant) => {
        const sVariant = this.activeSizeVariants.includes(cVariant)
        return !!sVariant
      })
    },
    onOptionsChange({ attributeCode, attributeId, option }) {
      /* istanbul ignore else */
      // if (attributeCode === 'color') {
      //   this.activeColorVariants = option.variants
      // }

      /* istanbul ignore else */
      if (attributeCode === 'size') {
        this.activeSizeVariants = option.variants
      }

      const variantId = this.getIntersectedVariantId()

      this.updateItem(variantId, this.item.quantity)
    },
    onQuantityChange(value) {
      this.updateItem(null, value)
    },
    updateItem(variant, quantity) {
      const payload = {
        cartId: this.$store.state.localStorage.cartId,
        itemId: this.item.id,
        body: { quantity }
      }

      /* istanbul ignore else */
      if (variant) {
        payload.body.variant = variant
      }

      this.$store.dispatch(`cart/${cartActionTypes.UPDATE_ITEM}`, payload)
    }
  }
}
</script>

<style lang="scss">
.order-item {
  display: flex;
  align-items: flex-start;

  &__remove {
    @include reset-button-styles();

    display: flex;
    flex: 0 0 30px;
    width: 30px;
    height: 30px;
    margin-top: -8px;
    transition: color 0.3s;
  }

  &__remove:hover &__icon,
  &__remove:focus &__icon,
  &__remove:active &__icon {
    color: $primary;
  }

  &__icon {
    flex: 0 0 15px;
    width: 15px;
    height: 15px;
    color: $text-color;
  }

  &__image-wrapper {
    position: relative;
    flex: 0 0 70px;
    height: 90px;
    margin-right: 14px;
  }

  &__image {
    max-width: 100%;
    max-height: 100%;
    object-fit: cover;
  }

  &__body {
    display: flex;
    flex-grow: 1;
    flex-wrap: wrap;
  }

  &__name {
    flex: 0 0 275px;
    max-width: 275px;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    font-size: 12px;
    line-height: 15px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__price-wrapper {
    margin-left: auto;
    text-align: right;
  }

  &__old-price {
    font-size: 11px;
    line-height: 14px;
    text-decoration-line: line-through;
    color: $text-color;
  }

  &__price {
    font-size: 14px;
    font-weight: 500;
    line-height: 18px;
    color: $primary;
  }

  &__code {
    width: 100%;
    margin-bottom: 17px;
    font-size: 12px;
    line-height: 15px;
    color: $gray-dark-1;
  }

  &__colors,
  &__sizes {
    align-self: center;
    margin-bottom: 2px;
  }

  &__colors {
    margin-right: 23px;
  }

  &__counter {
    margin-left: auto;
  }

  @include media-breakpoint-down(xl) {
    &__colors,
    &__sizes {
      margin-bottom: 4px;
    }
  }

  @include media-breakpoint-down(lg) {
    &__remove {
      justify-content: center;
      padding-right: 5px;
      margin-top: -9px;
    }

    &__icon {
      width: 13px;
      height: 13px;
    }

    &__image-wrapper {
      margin-right: 10px;
    }

    &__body {
      margin-top: -3px;
    }

    &__name {
      display: -webkit-box;
      flex: 0 0 170px;
      max-width: 170px;
    }

    &__code {
      margin-bottom: 19px;
      font-size: 11px;
      line-height: 17px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 0;
    }

    &__colors {
      margin-right: 15px;
    }
  }

  @include media-breakpoint-down(md) {
    &__remove {
      justify-content: flex-start;
    }

    &__image-wrapper {
      margin-right: 15px;
    }

    &__body {
      margin-top: 0;
    }

    &__name {
      flex: 0 0 275px;
      max-width: 275px;
    }

    &__code {
      margin-bottom: 15px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 4px;
    }

    &__colors {
      margin-right: 25px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__remove {
      flex: 0 0 23px;
      width: 23px;
      padding-right: 0;
    }

    &__image-wrapper {
      margin-right: 12px;
    }

    &__body {
      margin-top: -3px;
    }

    &__name {
      flex: 0 0 125px;
      max-width: 125px;
    }

    &__code {
      margin-bottom: 23px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 1px;
    }

    &__colors {
      margin-right: 10px;
    }
  }
}
</style>
