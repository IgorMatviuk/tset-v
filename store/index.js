import { actionTypes as catalogActionTypes } from '@/store/catalog'

export const mutationTypes = {
  SET_LANG: 'SET_LANG',
  SET_MENUS: 'SET_MENUS',
  SET_PAGE_TYPE: 'SET_PAGE_TYPE',
  SET_BREADCRUMB: 'SET_BREADCRUMB',
  SET_CONTACTS: 'SET_CONTACTS',
  SET_GLOBALS: 'SET_GLOBALS'
}

export const actionTypes = {
  GET_MENUS: 'GET_MENUS',
  GET_CONTACTS: 'GET_CONTACTS',
  GET_BANNER: 'GET_BANNER',
  GET_CLIENT_LOCATION: 'GET_CLIENT_LOCATION'
}

export const state = () => {
  return {
    locale: 'ru',
    locales: ['uk', 'ru'],
    fallbackLocale: 'ru',
    pageType: null,
    breadcrumb: [],
    menus: [],
    globals: {
      countryCode: 'UA',
      banner: null
    },
    contacts: {
      phone: '0 800 332 512',
      email: 'office@virna.com.ua',
      schedule: 'с 9:00 - 18:00',
      socialLinks: [
        { name: 'instagram', url: 'https://www.instagram.com/' },
        { name: 'facebook', url: 'https://facebook.com/' },
        { name: 'youtube', url: 'https://www.youtube.com/' }
      ]
    }
  }
}

export const mutations = {
  [mutationTypes.SET_LANG](state, locale) {
    state.locale = locale
  },
  [mutationTypes.SET_MENUS](state, menus) {
    state.menus = menus
  },
  [mutationTypes.SET_PAGE_TYPE](state, pageType) {
    state.pageType = pageType
  },
  [mutationTypes.SET_BREADCRUMB](state, breadcrumb) {
    state.breadcrumb = breadcrumb
  },
  [mutationTypes.SET_CONTACTS](state, contacts) {
    state.contacts = contacts
  },
  [mutationTypes.SET_GLOBALS](state, { key, value }) {
    state.globals[key] = value
  }
}

export const actions = {
  async [actionTypes.GET_MENUS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: menus } = await this.$axios.$get(`${STAGE_API_URL}/api/menus`)
    commit(mutationTypes.SET_MENUS, menus)
  },
  async [actionTypes.GET_CONTACTS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: contacts } = await this.$axios.$get(
      `${STAGE_API_URL}/api/contacts`
    )
    commit(mutationTypes.SET_CONTACTS, contacts)
  },
  async [actionTypes.GET_BANNER]({ commit }) {
    const { data: banner } = await this.$axios.$get('/api/banner')
    const defaultBanner = {
      key: 'banner',
      value: banner
    }

    commit(mutationTypes.SET_GLOBALS, defaultBanner)
  },
  async [actionTypes.GET_CLIENT_LOCATION]({ commit }) {
    const locationString = await this.$axios.$get('https://ip2c.org/s')
    const [, countryCode] = locationString.split(';')
    const defaultCountry = {
      key: 'countryCode',
      value: countryCode
    }

    commit(mutationTypes.SET_GLOBALS, defaultCountry)
  },
  async nuxtServerInit({ dispatch }) {
    await Promise.all([
      dispatch(actionTypes.GET_MENUS),
      dispatch(actionTypes.GET_CONTACTS),
      dispatch(`catalog/${catalogActionTypes.GET_CATEGORIES}`)
    ])
  }
}

/**
 * Get menum by type
 * @param {object} state
 * @param {('header'|'footer'|'catalog'|'promo')} type
 */
function getMenuByType(state, type) {
  const { menus } = state
  return menus.find((menu) => menu.menu_type === type)
}

export const getters = {
  mobileMenu(state) {
    return getMenuByType(state, 'mobile')
  },
  desktopMenu(state) {
    return getMenuByType(state, 'header')
  },
  promoMenu(state) {
    return getMenuByType(state, 'promo')
  },
  topMenu(state) {
    return getMenuByType(state, 'top')
  },
  catalogMenu(state) {
    return getMenuByType(state, 'catalog')
  },
  footerMenu(state) {
    return getMenuByType(state, 'footer')
  }
}
