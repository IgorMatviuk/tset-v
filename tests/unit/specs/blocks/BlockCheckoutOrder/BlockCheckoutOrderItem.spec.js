import { mount } from '@vue/test-utils'
import _cloneDeep from 'lodash.clonedeep'

import BlockCheckoutOrderItem from '@/components/blocks/BlockCheckoutOrder/BlockCheckoutOrderItem'

import { actionTypes as cartActionTypes } from '@/store/cart'

import productAttrs from '@/api/factories/product'
import createFactory from '@/api/helpers/factory'

describe('BlockCheckoutOrderItem', () => {
  const updateAction = `cart/${cartActionTypes.UPDATE_ITEM}`
  const removeAction = `cart/${cartActionTypes.REMOVE_ITEM}`

  let cartItems
  let regularItem
  let discountItem
  let mockStore

  function createOrderItem(item) {
    const wrapper = mount(BlockCheckoutOrderItem, {
      propsData: { item, page: 'checkout', index: 0 },
      mocks: {
        $route: {
          name: 'checkoutPage'
        },
        $t: (str) => str,
        $i18n: {
          path: (str) => str
        },
        $store: mockStore
      },
      stubs: ['svg-icon', 'NuxtLink', 'BPopover']
    })

    return wrapper
  }

  beforeEach(() => {
    cartItems = createFactory(productAttrs, 4).map((product, index) => {
      const cartItem = _cloneDeep(product)

      cartItem.quantity = index + 1

      cartItem.attributes.forEach((attribute) => {
        attribute.selected_option = attribute.options[0]
      })

      return cartItem
    })

    const regularProduct = cartItems.find((item) => !item.discount_price)
    regularItem = {
      id: 1,
      product: regularProduct,
      variant_id: regularProduct.variants[0].id,
      quantity: 1
    }

    const discountProduct = cartItems.find((item) => item.discount_price)
    discountItem = {
      id: 2,
      product: discountProduct,
      variant_id: discountProduct.variants[0].id,
      quantity: 2
    }

    mockStore = {
      state: {
        pageType: null,
        localStorage: {
          cartId: 1
        }
      },
      dispatch: jest.fn()
    }
  })

  test('has valid initial structure (regular)', () => {
    const wrapper = createOrderItem(regularItem)
    expect(wrapper).toMatchSnapshot()
  })

  test('has valid initial structure (discount)', () => {
    const wrapper = createOrderItem(discountItem)
    expect(wrapper).toMatchSnapshot()
  })

  test('updates item quantity', () => {
    const wrapper = createOrderItem(regularItem)

    expect(wrapper.vm.item.quantity).toBe(regularItem.quantity)
    const increasedQtyValue = wrapper.vm.item.quantity + 1
    wrapper.vm.onQuantityChange(increasedQtyValue)

    const payload = {
      cartId: wrapper.vm.$store.state.localStorage.cartId,
      itemId: wrapper.vm.item.id,
      body: { quantity: increasedQtyValue }
    }

    expect(mockStore.dispatch).toBeCalledWith(updateAction, payload)
  })

  test('updates item options', () => {
    const wrapper = createOrderItem(regularItem)

    const byCode = (code) => (item) => item.code === code
    const nextActiveItem = (item) => !item.disabled

    const { attributes } = wrapper.vm.item.product

    // On size update
    const sizeAttribute = attributes.find(byCode('size'))
    const nextSizeOption = wrapper.vm
      .mapOptionsWithStatus('size')
      .find(nextActiveItem)

    wrapper.vm.onOptionsChange({
      attributeCode: 'size',
      attributeId: sizeAttribute.id,
      option: nextSizeOption
    })

    const sizeUpdatePayload = {
      cartId: wrapper.vm.$store.state.localStorage.cartId,
      itemId: wrapper.vm.item.id,
      body: {
        variant: wrapper.vm.getIntersectedVariantId(),
        quantity: wrapper.vm.item.quantity
      }
    }

    expect(mockStore.dispatch).toBeCalledWith(updateAction, sizeUpdatePayload)

    // On color update
    // const colorAttribute = attributes.find(byCode('color'))
    // const nextColorOption = wrapper.vm
    //   .mapOptionsWithStatus('color')
    //   .find(nextActiveItem)

    // wrapper.vm.onOptionsChange({
    //   attributeCode: 'color',
    //   attributeId: colorAttribute.id,
    //   option: nextColorOption
    // })

    // const colorUpdatePayload = {
    //   cartId: wrapper.vm.$store.state.localStorage.cartId,
    //   itemId: wrapper.vm.item.id,
    //   body: {
    //     variant: wrapper.vm.getIntersectedVariantId(),
    //     quantity: wrapper.vm.item.quantity
    //   }
    // }

    // expect(mockStore.dispatch).toBeCalledWith(updateAction, colorUpdatePayload)
  })

  test('removes item from cart', () => {
    const wrapper = createOrderItem(regularItem)

    const { id: itemId } = wrapper.vm.item
    const { cartId } = wrapper.vm.$store.state.localStorage

    wrapper.find('.order-item__remove').trigger('click')

    expect(mockStore.dispatch).toBeCalledWith(removeAction, { cartId, itemId })
  })
})
