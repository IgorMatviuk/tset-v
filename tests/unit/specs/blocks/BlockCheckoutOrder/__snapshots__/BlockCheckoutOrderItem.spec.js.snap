// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockCheckoutOrderItem has valid initial structure (discount) 1`] = `
<div class="order-item"><button type="button" class="order-item__remove">
    <svg-icon-stub name="trash" width="15" height="15" class="order-item__icon"></svg-icon-stub>
  </button>
  <div class="order-item__image-wrapper">
    <div class="img-loader">
      <picture>
        <source data-srcset="/img/previews/preview-1-small.webp" srcset="/img/stub.png" type="image/webp"> <img data-src="/img/previews/preview-1-small.jpg" src="/img/stub.png" alt="Брюки кожаные с карманами" title="Брюки кожаные с карманами" loading="lazy" data-object-fit="cover" class="order-item__image lazyload"></picture>
    </div>
  </div>
  <div class="order-item__body">
    <nuxtlink-stub to="strujashheesja-platje-s-cvetochnym-printom-base-variant-1-7-0" class="order-item__name">
      Брюки кожаные с карманами
    </nuxtlink-stub>
    <div class="order-item__price-wrapper">
      <div class="order-item__price">
        750 грн
      </div>
      <div class="order-item__old-price">
        1500 грн
      </div>
    </div>
    <div class="order-item__code">
      productPage.sku: dress-base-variant-1-7
    </div>
    <div id="orderSize_2_0" class="options-select order-item__sizes">
      <div class="options-select__active-option options-select__active-option--size">
        XS
        <svg-icon-stub name="chevron" width="10" height="10" class="options-select__icon"></svg-icon-stub>
      </div>
      <bpopover-stub target="orderSize_2_0" placement="bottom">
        <div class="options-select__popover">
          <div class="options-select__option options-select__option--size options-select__option--active">
            XS
          </div>
          <div class="options-select__option options-select__option--size">
            S
          </div>
          <div class="options-select__option options-select__option--size options-select__option--disabled">
            M
          </div>
        </div>
      </bpopover-stub>
    </div>
    <div class="counter order-item__counter"><button type="button" class="counter__toggle counter__toggle--decrease">
        <svg-icon-stub name="minus" height="1" width="17" class="counter__icon"></svg-icon-stub>
      </button> <input type="text" class="counter__input"> <button type="button" class="counter__toggle counter__toggle--increase">
        <svg-icon-stub name="plus" height="17" width="17" class="counter__icon"></svg-icon-stub>
      </button></div>
  </div>
</div>
`;

exports[`BlockCheckoutOrderItem has valid initial structure (regular) 1`] = `
<div class="order-item"><button type="button" class="order-item__remove">
    <svg-icon-stub name="trash" width="15" height="15" class="order-item__icon"></svg-icon-stub>
  </button>
  <div class="order-item__image-wrapper">
    <div class="img-loader">
      <picture>
        <source data-srcset="/img/previews/preview-2-small.webp" srcset="/img/stub.png" type="image/webp"> <img data-src="/img/previews/preview-2-small.jpg" src="/img/stub.png" alt="Хлопковое платье" title="Хлопковое платье" loading="lazy" data-object-fit="cover" class="order-item__image lazyload"></picture>
    </div>
  </div>
  <div class="order-item__body">
    <nuxtlink-stub to="strujashheesja-platje-s-cvetochnym-printom-base-variant-1-7-1" class="order-item__name">
      Хлопковое платье
    </nuxtlink-stub>
    <div class="order-item__price-wrapper">
      <div class="order-item__price">
        1500 грн
      </div>
    </div>
    <div class="order-item__code">
      productPage.sku: dress-base-variant-1-7
    </div>
    <div id="orderSize_1_0" class="options-select order-item__sizes">
      <div class="options-select__active-option options-select__active-option--size">
        XS
        <svg-icon-stub name="chevron" width="10" height="10" class="options-select__icon"></svg-icon-stub>
      </div>
      <bpopover-stub target="orderSize_1_0" placement="bottom">
        <div class="options-select__popover">
          <div class="options-select__option options-select__option--size options-select__option--active">
            XS
          </div>
          <div class="options-select__option options-select__option--size">
            S
          </div>
          <div class="options-select__option options-select__option--size options-select__option--disabled">
            M
          </div>
        </div>
      </bpopover-stub>
    </div>
    <div class="counter order-item__counter"><button type="button" class="counter__toggle counter__toggle--decrease">
        <svg-icon-stub name="minus" height="1" width="17" class="counter__icon"></svg-icon-stub>
      </button> <input type="text" class="counter__input"> <button type="button" class="counter__toggle counter__toggle--increase">
        <svg-icon-stub name="plus" height="17" width="17" class="counter__icon"></svg-icon-stub>
      </button></div>
  </div>
</div>
`;
